src/task_types/resmoke_tasks.rs

```diff
@@ -396,13 +407,23 @@ impl GenResmokeTaskServiceImpl {
         task_stats: &TaskRuntimeHistory,
         multiversion_name: Option<&str>,
         multiversion_tags: Option<String>,
+        preferred_stat: PreferredStatForSplitTask,
     ) -> Result<Vec<SubSuite>> {
+        fn select_stat(average_runtime: f64, max_duration: f64, preferred_stat_: &PreferredStatForSplitTask) -> f64 {
```

Comment:
I think it would be cleaner to create a new `trait` called `SmartStatSelector` with a default implementation for a method called `select_stat` which takes in a single param of type `ResmokeGenParams` (ie: check [this](https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations) link).

Then we can require `HookRuntimeHistory` and `TestRuntimeHistory` to implement this trait (but use the default implementation) & directly use the `select_stat` method from the struct.

The `ResmokeGenParams` parameter can provide the necessary information to determine which stat to use in the `select_stat` method. 

^ A simplistic way is to determine which stat is to use the [task_name](https://github.com/mongodb/mongo-task-generator/blob/7a9e4678ddfe48f8b73bf86961a8c6370d41deab/src/task_types/resmoke_tasks.rs#L48) and see if it starts with `benchmarks_*`, but a better way would maybe be to extract some additional information from the evergreen config when [constructing the `ResmokeGenParams` struct](https://github.com/mongodb/mongo-task-generator/blob/7a9e4678ddfe48f8b73bf86961a8c6370d41deab/src/services/config_extraction.rs#L218
) itself and store it as an attirbute. Then, use that attribute to determine whether to use the max or avg stat.

Reply:
^^ Btw, we have used evergreen task tags in the past for this. So we could use a new tag (ie: `use_max_time_for_split` or similar). So, any `*_gen` task with that tag will use the `max_time` stat instead of the default `avg_time` stat.

Here's a PR comment where I did something similar: https://github.com/mongodb/mongo-task-generator/pull/59/files#r1262557113

Reply:
Per offline conversation: refactored a fair bit.